dex/networks/eth/txdata.go

```diff
@@ -62,32 +62,33 @@ func ParseInitiateData(calldata []byte) ([]ETHSwapInitiation, error) {
 // contract with extra data. It will error if the call data does not call
 // redeem with expected argument types. It returns the secret and secret hash
 // in that order.
-func ParseRedeemData(calldata []byte) (secret [32]byte, secretHash [32]byte, err error) {
-	fail := func(err error) ([32]byte, [32]byte, error) {
-		return [32]byte{}, [32]byte{}, err
-	}
+func ParseRedeemData(calldata []byte) ([]ETHSwapRedemption, error) {
 	decoded, err := parseCallData(calldata, ETHSwapABI)
 	if err != nil {
-		return fail(fmt.Errorf("unable to parse call data: %v", err))
+		return nil, fmt.Errorf("unable to parse call data: %v", err)
 	}
 	if decoded.name != redeemFuncName {
-		return fail(fmt.Errorf("expected %v function but got %v", redeemFuncName, decoded.name))
+		return nil, fmt.Errorf("expected %v function but got %v", redeemFuncName, decoded.name)
 	}
 	args := decoded.inputs
 	// Any difference in number of args and types than what we expect
 	// should be caught by parseCallData, but checking again anyway.
 	//
 	// TODO: If any of the checks prove redundant, remove them.
 	if len(args) != numRedeemArgs {
-		return fail(fmt.Errorf("expected %v redeem args but got %v", numRedeemArgs, len(args)))
+		return nil, fmt.Errorf("expected %v redeem args but got %v", numRedeemArgs, len(args))
 	}
-	secret, ok := args[0].value.([32]byte)
+	redemptions, ok := args[0].value.([]struct {
+		Secret     [32]byte `json:"secret"`
+		SecretHash [32]byte `json:"secretHash"`
+	})
```

Comment:
It super sucks that `ETHSwapRedemption` (generated code) does not tag the struct fields.
I wonder if we should make an issue with geth.  Maybe they'd tell us to do something differently.

Reply:
The code we are using to examine the txdata is not even exported. I doubt they want us to do this. We are supposed to just use the contract I believe.

Reply:
I think they would say that too.  My note about a real audit involving isRedeemable before attempting the redeem is a required step we will need to take. https://github.com/decred/dcrdex/pull/1274#discussion_r754401945